src/gui/utils/plasmacolorscheme.cpp

```diff
@@ -0,0 +1,434 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "plasmacolorscheme.h"
+
+#include <stdexcept>
+
+#include <QDir>
+#include <QApplication>
+#include <QSettings>
+#include <QStandardPaths>
+
+#include "colorutils.h"
+
+#include <QDebug>
+
+namespace {
+    static QString settingsGroupName(PlasmaColorScheme::ColorSet set)
+    {
+        switch(set) {
+        case PlasmaColorScheme::View:
+            return QLatin1String("Colors:View");
+        case PlasmaColorScheme::Window:
+            return QLatin1String("Colors:Window");
+        case PlasmaColorScheme::Button:
+            return QLatin1String("Colors:Button");
+        case PlasmaColorScheme::Selection:
+            return QLatin1String("Colors:Selection");
+        case PlasmaColorScheme::Tooltip:
+            return QLatin1String("Colors:Tooltip");
+        default:
+            throw std::logic_error("Unexpected value of PlasmaColorScheme::ColorSet");
+        }
+    }
+
+    QRgb decodeKDEColor(const QVariant &v, bool &ok)
+    {
+        QVariantList components(v.toList());
+        if (components.size() == 3) {
+            bool rOk, gOk, bOk;
+            const int r = components[0].toString().toInt(&rOk);
+            const int g = components[1].toString().toInt(&gOk);
+            const int b = components[2].toString().toInt(&bOk);
+            if (rOk && gOk && bOk) {
+                ok = true;
+                return qRgb(r, g, b);
+            }
+        }
+        ok = false;
+        return qRgb(0, 0, 0);
+    }
+
+    PlasmaColorScheme::Effect readColorEffect(QSettings &settings, const QString &name, bool &ok)
+    {
+        using Effect = PlasmaColorScheme::Effect;
+        Effect res;
+        bool allReadsOK = true;
+        bool singleReadOK = false;
+        settings.beginGroup(QLatin1String("ColorEffects:") + name);
+
+        res.changeSelectionColor = settings.value(QLatin1String("ChangeSelectionColor"), false).toBool();
+
+        res.color = decodeKDEColor(settings.value(QLatin1String("Color")), singleReadOK);
+        allReadsOK &= singleReadOK;
```

Comment:
I don't know why this was done, apparently, for backward compatibility. IMO, implicit conversion between boolean and numeric types is evil.
>I better refactor this fragment to return early after the first read error.

It would be better independently of the discussed issue unless you need to read it all in any case.